src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,153 @@
+* If the primary lock disappeared with nothing left, it means the transaction has been rolled back. Then the row `R1`'s stale lock should also be rolled back. Then `T1` can continue.
+* If the primary lock exists but it's too old (we can set a TTL to the lock), we can believe that the transaction crashed before committing or rolling back. We can rollback it and continue `T1`.
+* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
+
+## Percolator in TiKV
+
+TiKV is a key-value storage, and each key-value pair can be regarded as a row in Percolator.
+
+TiKV internally uses RocksDB, a key-value storage engine library, to persist data to disk. RocksDB's atomic write batch and TiKV's transaction scheduler make it atomic to read and write a single user key, which is a requirement of Percolator.
+
+RocksDB provides a feature named *Column Family* (we will abbreviate it as *CF*). An instance of RocksDB may have multiple CFs, and each CF is a separated key namespace and has its own LSM-Tree. However different CFs in the same RocksDB instance uses a common WAL, providing the ability to write to different CFs atomically.
+
+We divide a RocksDB to three CFs: `CF_DEFAULT`, `CF_LOCK` and `CF_WRITE`, which corresponds to Percolator's `data` column, `lock` column and `write` column respectively. There's an extra CF named `CF_RAFT` used to save some metadata of Raft, but that's out of our topic now. The `notify` and `ack_O` columns are not present in TiKV, because for now TiKV doesn't need the ability of incremental processing.
+
+Then, we need to represent different versions of a key. We can simply compound a key and a timestamp as an internal key, and use the internal key in RocksDB. But since a key can have at most one lock at a time, so we don't need to add a timestamp to the key in `CF_LOCK`. Then the contents of each CFs are:
+
+* `CF_DEFAULT`: `(key, start_ts)` -> `value`
+* `CF_LOCK`: `key` -> `lock_info`
+* `CF_WRITE`: `(key, commit_ts)` -> `write_info`
+
+Our technical to compose user keys and timestamps together is:
+
+1. Encode the user key to [memcomparable](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format)
+2. Bitwise invert the timestamp (which is an unsigned int64) and encode it to big-endian bytes
+3. Append the encoded timestamp to the encoded key.
+
+For example, the key `"key1"` and the timestamp `3` will be encoded as `"key1\x00\x00\x00\x00\xfb\xff\xff\xff\xff\xff\xff\xff\xfe"`, where the first 9 bytes is the memcomparable-encoded key and the remaining 8 bytes is the inverted timestamp in big-endian. In this way, our different versions of the same key are always adjacent in RocksDB, and for each key, newer versions are always in front of older versions.
```

Comment:
```suggestion
For example, key `"key1"` and timestamp `3` will be encoded as `"key1\x00\x00\x00\x00\xfb\xff\xff\xff\xff\xff\xff\xff\xfe"`, where the first 9 bytes is the memcomparable-encoded key and the remaining 8 bytes is the inverted timestamp in big-endian. In this way, different versions of the same key are always adjacent in RocksDB; and for each key, newer versions are always before older ones.
```

Reply:
"... are always positionally before older ones"?

Reply:
I think the context is enough to indicate "positionally".

Reply:
I think the context is enough to indicate "positionally"